File: ios/Classes/Mapper.swift
import Flutter
import TwilioConversationsClient

public class Mapper {
    public static func conversationsClientToDict(_ client: TwilioConversationsClient?) -> [String: Any] {
        return [
            "conversations": conversationsToDict(client?.myConversations()) as Any,
            "myIdentity": client?.user?.identity as Any,
            "connectionState": clientConnectionStateToString(client?.connectionState),
            "isReachabilityEnabled": client?.isReachabilityEnabled() as Any
        ]
    }

    public static func conversationsClientToPigeon(_ client: TwilioConversationsClient?) -> TWCONConversationClientData? {
        guard let client = client else {
            return nil
        }
        let result = TWCONConversationClientData()
        result.myIdentity = client.user?.identity
        result.connectionState = clientConnectionStateToString(client.connectionState)
        result.isReachabilityEnabled = NSNumber(value: client.isReachabilityEnabled())
        return result
    }

    public static func conversationsToDict(_ conversations: [TCHConversation]?) -> [[String: Any]?]? {
        return conversations?.map { conversationToDict($0) }
    }
    
    public static func conversationsList(_ conversations: [TCHConversation]?) -> [TWCONConversationData]? {
        return conversations?.compactMap { conversationToPigeon($0) }
    }
    
    public static func conversationToDict(_ conversation: TCHConversation?) -> [String: Any]? {
        guard let conversation = conversation,
              let sid = conversation.sid else {
            return nil
        }
        
        if !SwiftTwilioConversationsPlugin.conversationListeners.keys.contains(sid) {
            SwiftTwilioConversationsPlugin.debug("Creating ConversationListener for conversation: '\(String(describing: conversation.sid))'")
            SwiftTwilioConversationsPlugin.conversationListeners[sid] = ConversationListener(sid)
            conversation.delegate = SwiftTwilioConversationsPlugin.conversationListeners[sid]
        }
        
        return [
            "attributes": attributesToDict(conversation.attributes()) as Any,
            "createdBy": conversation.createdBy as Any,
            "dateCreated": dateToString(conversation.dateCreatedAsDate) as Any,
            "dateUpdated": dateToString(conversation.dateUpdatedAsDate) as Any,
            "friendlyName": conversation.friendlyName as Any,
            "lastMessageDate": dateToString(conversation.lastMessageDate) as Any,
            "lastMessageIndex": conversation.lastMessageIndex as Any,
            "lastReadMessageIndex": conversation.lastReadMessageIndex as Any,
            "sid": sid,
            "status": conversationStatusToString(conversation.status),
            "synchronizationStatus": conversationSynchronizationStatusToString(conversation.synchronizationStatus),
            "uniqueName": conversation.uniqueName as Any
        ]
    }
    
    public static func conversationToPigeon(_ conversation: TCHConversation?) -> TWCONConversationData? {
        guard let conversation = conversation,
              let sid = conversation.sid else {
            return nil
        }
        
        if !SwiftTwilioConversationsPlugin.conversationListeners.keys.contains(sid) {
            SwiftTwilioConversationsPlugin.debug("Creating ConversationListener for conversation: '\(String(describing: conversation.sid))'")
            SwiftTwilioConversationsPlugin.conversationListeners[sid] = ConversationListener(sid)
            conversation.delegate = SwiftTwilioConversationsPlugin.conversationListeners[sid]
        }
        
        let result = TWCONConversationData()
        result.attributes = attributesToPigeon(conversation.attributes())
        result.createdBy = conversation.createdBy
        result.dateCreated = dateToString(conversation.dateCreatedAsDate)
        result.dateUpdated = dateToString(conversation.dateUpdatedAsDate)
        result.friendlyName = conversation.friendlyName
        result.lastMessageDate = dateToString(conversation.lastMessageDate)
        result.lastMessageIndex = conversation.lastMessageIndex
        result.lastReadMessageIndex = conversation.lastReadMessageIndex
        result.sid = sid
        result.status = conversationStatusToString(conversation.status)
        result.synchronizationStatus = conversationSynchronizationStatusToString(conversation.synchronizationStatus)
        result.uniqueName = conversation.uniqueName
        return result
    }

    public static func messageToPigeon(_ message: TCHMessage, conversationSid: String?) -> TWCONMessageData {
        let result = TWCONMessageData()
        result.sid = message.sid
        result.author = message.author
        result.dateCreated = message.dateCreated
        result.dateUpdated = message.dateUpdated
        result.lastUpdatedBy = message.lastUpdatedBy
        result.messageBody = message.body
        result.conversationSid = conversationSid
        result.participantSid = message.participantSid
//        result.participant = participantToDict(message.participant, conversationSid: conversationSid)
        result.messageIndex = message.index
        result.type = messageTypeToString(message.messageType)
        result.hasMedia = NSNumber(value: message.hasMedia())
        result.media = mediaToPigeon(message, conversationSid)
        result.attributes = attributesToPigeon(message.attributes())
        return result
    }

    public static func messageToDict(_ message: TCHMessage, conversationSid: String?) -> [String: Any?] {
        return [
            "sid": message.sid,
            "author": message.author,
            "dateCreated": message.dateCreated,
            "dateUpdated": message.dateUpdated,
            "lastUpdatedBy": message.lastUpdatedBy,
            "messageBody": message.body,
            "conversationSid": conversationSid,
            "participantSid": message.participantSid,
            "participant": participantToDict(message.participant, conversationSid: conversationSid),
            "messageIndex": message.index,
            "type": messageTypeToString(message.messageType),
            "hasMedia": message.hasMedia(),
            "media": mediaToDict(message, conversationSid),
            "attributes": attributesToDict(message.attributes())
        ]
    }
    
    public static func participantToDict(_ participant: TCHParticipant?, conversationSid: String?) -> [String: Any?]? {
        guard let participant = participant else {
            return nil
        }
        return [
            "sid": participant.sid,
            "conversationSid": conversationSid,
            "lastReadMessageIndex": participant.lastReadMessageIndex,
            "lastReadTimestamp": participant.lastReadTimestamp,
            "dateCreated": participant.dateCreated,
            "dateUpdated": participant.dateUpdated,
            "identity": participant.identity,
            "type": participantTypeToString(participant.type)
        ]
    }
    
    public static func participantToPigeon(_ participant: TCHParticipant?, conversationSid: String?) -> TWCONParticipantData? {
        guard let participant = participant else {
            return nil
        }
        
        let result = TWCONParticipantData()
        result.sid = participant.sid
        result.conversationSid = conversationSid
        result.lastReadMessageIndex = participant.lastReadMessageIndex
        result.lastReadTimestamp = participant.lastReadTimestamp
        result.dateCreated = participant.dateCreated
        result.dateUpdated = participant.dateUpdated
        result.identity = participant.identity
        result.type = participantTypeToString(participant.type)
        result.attributes = attributesToPigeon(participant.attributes())
        return result
    }

    public static func userToDict(_ user: TCHUser?) -> [String: Any]? {
        guard let user = user else {
            return nil
        }
        return [
            "friendlyName": user.friendlyName as Any,
            "attributes": attributesToDict(user.attributes()) as Any,
            "identity": user.identity as Any,
            "isOnline": user.isOnline(),
            "isNotifiable": user.isNotifiable(),
            "isSubscribed": user.isSubscribed()
        ]
    }
    
    public static func userToPigeon(_ user: TCHUser?) -> TWCONUserData? {
        guard let user = user else {
            return nil
        }
        let result = TWCONUserData()
        result.friendlyName = user.friendlyName
        result.attributes = attributesToPigeon(user.attributes())
        result.identity = user.identity
        result.isOnline = NSNumber(value: user.isOnline())
        result.isNotifiable = NSNumber(value: user.isNotifiable())
        result.isSubscribed = NSNumber(value: user.isSubscribed())
        return result
    }
    
    public static func mediaToDict(_ message: TCHMessage, _ conversationSid: String?) -> [String: Any?]? {
        if !message.hasMedia() {
            return nil
        }
        return [
            "sid": message.mediaSid,
            "fileName": message.mediaFilename,
            "type": message.mediaType,
            "size": message.mediaSize,
            "conversationSid": conversationSid,
            "messageIndex": message.index,
            "messageSid": message.sid
        ]
    }
    
    public static func mediaToPigeon(_ message: TCHMessage, _ conversationSid: String?) -> TWCONMessageMediaData? {
        if !message.hasMedia() {
            return nil
        }
        
        let result = TWCONMessageMediaData()
        result.sid = message.mediaSid
        result.fileName = message.mediaFilename
        result.type = message.mediaType
        result.size = NSNumber(value: message.mediaSize)
        result.conversationSid = conversationSid
        result.messageIndex = message.index
        result.messageSid = message.sid
        return result
    }
    
    public static func attributesToDict(_ attributes: TCHJsonAttributes?) -> [String: Any?]? {
        if let attr = attributes as TCHJsonAttributes? {
            if attr.isNull {
                return [
                    "type": "NULL",
                    "data": nil
                ]
            } else if attr.isNumber {
                return [
                    "type": "NUMBER",
                    "data": attr.number?.stringValue
                ]
            } else if attr.isArray {
                guard let jsonData = try? JSONSerialization.data(withJSONObject: attr.array as Any) else {
                    return nil
                }
                return [
                    "type": "ARRAY",
                    "data": String(data: jsonData, encoding: String.Encoding.utf8)
                ]
            } else if attr.isString {
                return [
                    "type": "STRING",
                    "data": attr.string
                ]
            } else if attr.isDictionary {
                guard let jsonData = try? JSONSerialization.data(withJSONObject: attr.dictionary as Any) else {
                    return nil
                }
                return [
                    "type": "OBJECT",
                    "data": String(data: jsonData, encoding: String.Encoding.utf8)
                ]
            }
        }
        return nil
    }
    
    public static func attributesToPigeon(_ attributes: TCHJsonAttributes?) -> TWCONAttributesData? {
        let result = TWCONAttributesData()
        result.type = "NULL"
        result.data = nil
        
        if let attr = attributes as TCHJsonAttributes? {
            if attr.isNumber {
                result.type = "NUMBER"
                result.data = attr.number?.stringValue
            } else if attr.isArray {
                guard let jsonData = try? JSONSerialization.data(withJSONObject: attr.array as Any) else {
                    return result
                }
                result.type = "ARRAY"
                result.data = String(data: jsonData, encoding: String.Encoding.utf8)
            } else if attr.isString {
                result.type = "STRING"
                result.data = attr.string
            } else if attr.isDictionary {
                guard let jsonData = try? JSONSerialization.data(withJSONObject: attr.dictionary as Any) else {
                    return result
                }
                result.type = "OBJECT"
                result.data = String(data: jsonData, encoding: String.Encoding.utf8)
            }
        }
        return result
    }
    
    public static func dictToAttributes(_ dict: [String: Any?]) -> TCHJsonAttributes {
        return TCHJsonAttributes.init(dictionary: dict as [AnyHashable: Any])
    }
    
    public static func errorToDict(_ error: Error?) -> [String: Any?]? {
        if let error = error as NSError? {
            return [
                "code": error.code,
                "message": error.description
            ]
        }
        
        return nil
    }

    public static func errorToPigeon(_ error: TCHError) -> TWCONErrorInfoData {
        let errorInfoData = TWCONErrorInfoData()
        errorInfoData.code = NSNumber(value: error.code)
        errorInfoData.message = error.description
        return errorInfoData
    }

    public static func conversationStatusToString(_ conversationStatus: TCHConversationStatus) -> String {
        let conversationStatusString: String
        
        switch conversationStatus {
        case .joined:
            conversationStatusString = "JOINED"
        case .notParticipating:
            conversationStatusString = "NOT_PARTICIPATING"
        @unknown default:
            conversationStatusString = "UNKNOWN"
        }
        
        return conversationStatusString
    }
    
    public static func clientConnectionStateToString(_ connectionState: TCHClientConnectionState?) -> String {
        var connectionStateString: String = "UNKNOWN"
        if let connectionState = connectionState {
            switch connectionState {
            case .unknown:
                connectionStateString = "UNKNOWN"
            case .disconnected:
                connectionStateString = "DISCONNECTED"
            case .connected:
                connectionStateString = "CONNECTED"
            case .connecting:
                connectionStateString = "CONNECTING"
            case .denied:
                connectionStateString = "DENIED"
            case .error:
                connectionStateString = "ERROR"
            case .fatalError:
                connectionStateString = "FATAL_ERROR"
            default:
                connectionStateString = "UNKNOWN"
            }
        }
        
        return connectionStateString
    }
    
    public static func clientSynchronizationStatusToString(_ syncStatus: TCHClientSynchronizationStatus?) -> String {
        var syncStateString: String = "UNKNOWN"
        if let syncStatus = syncStatus {
            switch syncStatus {
            case .started:
                syncStateString = "STARTED"
            case .completed:
                syncStateString = "COMPLETED"
            case .conversationsListCompleted:
                syncStateString = "CONVERSATIONS_COMPLETED"
            case .failed:
                syncStateString = "FAILED"
            @unknown default:
                syncStateString = "UNKNOWN"
            }
        }
        
        return syncStateString
    }
    
    public static func conversationSynchronizationStatusToString(_ syncStatus: TCHConversationSynchronizationStatus) -> String {
        let syncStatusString: String
        
        switch syncStatus {
        case .none:
            syncStatusString = "NONE"
        case .identifier:
            syncStatusString = "IDENTIFIER"
        case .metadata:
            syncStatusString = "METADATA"
        case .all:
            syncStatusString = "ALL"
        case .failed:
            syncStatusString = "FAILED"
        @unknown default:
            syncStatusString = "UNKNOWN"
        }
        
        return syncStatusString
    }
    
    public static func conversationUpdateToString(_ update: TCHConversationUpdate) -> String {
        switch update {
        case .attributes:
            return "ATTRIBUTES"
        case .friendlyName:
            return "FRIENDLY_NAME"
        case .lastMessage:
            return "LAST_MESSAGE"
        case .lastReadMessageIndex:
            return "LAST_READ_MESSAGE_INDEX"
        case .state:
            return "STATE"
        case .status:
            return "STATUS"
        case .uniqueName:
            return "UNIQUE_NAME"
        case .userNotificationLevel:
            return "NOTIFICATION_LEVEL"
        @unknown default:
            return "UNKNOWN"
        }
    }
    
    public static func participantTypeToString(_ participantType: TCHParticipantType) -> String {
        let participantTypeString: String
        
        switch participantType {
        case .chat:
            participantTypeString = "CHAT"
        case .other:
            participantTypeString = "OTHER"
        case .sms:
            participantTypeString = "SMS"
        case .unset:
            participantTypeString = "UNSET"
        case .whatsapp:
            participantTypeString = "WHATSAPP"
        @unknown default:
            participantTypeString = "UNKNOWN"
        }
        
        return participantTypeString
    }
    
    public static func messageTypeToString(_ messageType: TCHMessageType) -> String {
        let messageTypeString: String
        
        switch messageType {
        case .media:
            messageTypeString = "MEDIA"
        case .text:
            messageTypeString = "TEXT"
        @unknown default:
            messageTypeString = "UNKNOWN"
        }
        
        return messageTypeString
    }
    
    public static func messageUpdateToString(_ update: TCHMessageUpdate) -> String {
        let updateString: String
        
        switch update {
        case .attributes:
            updateString = "ATTRIBUTES"
        case .body:
            updateString = "BODY"
        case .deliveryReceipt:
            updateString = "DELIVERY_RECEIPT"
        case .subject:
            updateString = "SUBJECT"
        @unknown default:
            updateString = "UNKNOWN"
        }
        
        return updateString
    }
    
    public static func participantUpdateToString(_ update: TCHParticipantUpdate) -> String {
        let updateString: String
        
        switch update {
        case .attributes:
            updateString = "ATTRIBUTES"
        case .lastReadMessageIndex:
            updateString = "LAST_READ_MESSAGE_INDEX"
        case .lastReadTimestamp:
            updateString = "LAST_READ_TIMESTAMP"
        @unknown default:
            updateString = "UNKNOWN"
        }
        
        return updateString
    }
    
    public static func userUpdateToString(_ update: TCHUserUpdate) -> String {
        switch update {
        case .friendlyName:
            return "FRIENDLY_NAME"
        case .attributes:
            return "ATTRIBUTES"
        case .reachabilityOnline:
            return "REACHABILITY_ONLINE"
        case .reachabilityNotifiable:
            return "REACHABILITY_NOTIFIABLE"
        @unknown default:
            return "UNKNOWN"
        }
    }
    
    public static func dateToString(_ date: Date?) -> String? {
        if let date = date {
            let formatter = DateFormatter()
            formatter.dateFormat = "yyyy-MM-dd HH:mm:ss ZZZ"
            return formatter.string(from: date)
        }
        return nil
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/Mapper.kt
package twilio.flutter.twilio_conversations

import com.twilio.conversations.*
import org.json.JSONArray
import org.json.JSONObject
import twilio.flutter.twilio_conversations.listeners.ConversationListener
import java.text.SimpleDateFormat
import java.util.*

object Mapper {
    //TODO go through all of the mappers in iOS, Android, and Dart, to make sure
    // they are consistent
    fun conversationsClientToPigeon(client: ConversationsClient): Api.ConversationClientData {
        val result = Api.ConversationClientData()
        result.myIdentity = client.myIdentity
        result.connectionState = client.connectionState.toString()
        result.isReachabilityEnabled = client.isReachabilityEnabled
        return result
    }

    fun attributesToMap(attributes: Attributes): Map<String, Any> {
        return mapOf(
                "type" to attributes.type.toString(),
                "data" to attributes.toString()
        )
    }

    fun listToJSONArray(list: List<Any>): JSONArray {
        val result = JSONArray()
        list.forEach {
            if (it is Map<*, *>) {
                result.put(mapToJSONObject(it as Map<String, Any>))
            } else if (it is List<*>) {
                result.put(listToJSONArray(it as List<Any>))
            } else {
                result.put(it)
            }
        }
        return result
    }

    fun mapToJSONObject(map: Map<String, Any>?): JSONObject? {
        if (map == null) {
            return null
        }
        val result = JSONObject()
        map.keys.forEach {
            if (map[it] == null) {
                result.put(it, null)
            } else if (map[it] is Map<*, *>) {
                result.put(it, mapToJSONObject(map[it] as Map<String, Any>))
            } else if (map[it] is List<*>) {
                result.put(it, listToJSONArray(map[it] as List<Any>))
            } else {
                result.put(it, map[it])
            }
        }
        return result
    }

    fun attributesToPigeon(attributes: Attributes): Api.AttributesData {
        val result = Api.AttributesData()
        result.type = attributes.type.toString()
        result.data = attributes.toString()
        return result
    }

    fun pigeonToAttributes(pigeon: Api.AttributesData): Attributes {
        var result = Attributes()
        when (pigeon.type) {
            "NULL" ->
                result = Attributes()
            "BOOLEAN" ->
                result = Attributes(pigeon.data.toBoolean())
            "NUMBER" -> {
                val number: Number =
                    if (pigeon.data.contains('.')) pigeon.data.toFloat()
                    else pigeon.data.toInt()
                result = Attributes(number)
            }
            "STRING" ->
                result = Attributes(pigeon.data)
            "OBJECT" ->
                result = Attributes(JSONObject(pigeon.data))
            "ARRAY" ->
                result = Attributes(JSONArray(pigeon.data))
        }
        return result
    }

    fun conversationsListToPigeon(conversations: MutableList<Conversation>?): List<Api.ConversationData> {
        if (conversations == null) {
            return listOf()
        }
        return conversations.mapNotNull { conversationToPigeon(it) }
    }

    fun conversationToPigeon(conversation: Conversation?): Api.ConversationData? {
        if (conversation == null) return null

        // Setting flutter event listener for the given channel if one does not yet exist.
        if (conversation.sid != null && !TwilioConversationsPlugin.conversationListeners.containsKey(conversation.sid)) {
            TwilioConversationsPlugin.debug("Creating ConversationListener for conversation: '${conversation.sid}'")
            TwilioConversationsPlugin.conversationListeners[conversation.sid] = ConversationListener(conversation.sid)
            conversation.addListener(TwilioConversationsPlugin.conversationListeners[conversation.sid])
        }

        val result = Api.ConversationData()
        result.createdBy = conversation.createdBy
        result.dateCreated = dateToString(conversation.dateCreatedAsDate)
        result.dateUpdated = dateToString(conversation.dateUpdatedAsDate)
        result.friendlyName = conversation.friendlyName
        result.lastMessageDate = dateToString(conversation.lastMessageDate)
        result.lastReadMessageIndex =
            if (conversation.synchronizationStatus.isAtLeast(Conversation.SynchronizationStatus.METADATA))
                conversation.lastReadMessageIndex else null
        result.lastMessageIndex = conversation.lastMessageIndex
        result.sid = conversation.sid
        result.status = conversation.status.toString()
        result.synchronizationStatus = conversation.synchronizationStatus.toString()
        result.uniqueName = conversation.uniqueName

        result.attributes = attributesToPigeon(conversation.attributes)

        return result
    }

    fun messageToPigeon(message: Message): Api.MessageData {
        val result = Api.MessageData()

        result.sid = message.sid
        result.author = message.author
        result.dateCreated= dateToString(message.dateCreatedAsDate)
        result.dateUpdated = dateToString(message.dateUpdatedAsDate)
        result.lastUpdatedBy = message.lastUpdatedBy
        result.subject = message.subject
        result.messageBody = message.messageBody
        result.conversationSid = message.conversation.sid
        result.participantSid = message.participantSid
//        result.participant = participantToMap(message.participant)
        result.messageIndex = message.messageIndex
        result.type = message.type.toString()
        result.media = mediaToPigeon(message)
        result.hasMedia = message.hasMedia()
        result.attributes = attributesToPigeon(message.attributes)
        return result
    }

    fun conversationToMap(conversation: Conversation?): Map<String, Any?>? {
        if (conversation == null) return null

        // Setting flutter event listener for the given channel if one does not yet exist.
        if (conversation.sid != null && !TwilioConversationsPlugin.conversationListeners.containsKey(conversation.sid)) {
            TwilioConversationsPlugin.debug("Creating ConversationListener for conversation: '${conversation.sid}'")
            TwilioConversationsPlugin.conversationListeners[conversation.sid] = ConversationListener(conversation.sid)
            conversation.addListener(TwilioConversationsPlugin.conversationListeners[conversation.sid])
        }

        return mapOf(
            "attributes" to attributesToMap(conversation.attributes),
            "createdBy" to conversation.createdBy,
            "dateCreated" to dateToString(conversation.dateCreatedAsDate),
            "dateUpdated" to dateToString(conversation.dateUpdatedAsDate),
            "friendlyName" to conversation.friendlyName,
            "lastMessageDate" to dateToString(conversation.lastMessageDate),
            "lastReadMessageIndex" to
                    if (conversation.synchronizationStatus.isAtLeast(Conversation.SynchronizationStatus.METADATA))
                        conversation.lastReadMessageIndex else null,
            "lastMessageIndex" to conversation.lastMessageIndex,
            "sid" to conversation.sid,
            "status" to conversation.status.toString(),
            "synchronizationStatus" to conversation.synchronizationStatus.toString(),
            "uniqueName" to conversation.uniqueName
        )
    }

    fun messageToMap(message: Message): Map<String, Any?> {
        return mapOf(
                "sid" to message.sid,
                "author" to message.author,
                "dateCreated" to dateToString(message.dateCreatedAsDate),
                "dateUpdated" to dateToString(message.dateUpdatedAsDate),
                "lastUpdatedBy" to message.lastUpdatedBy,
                "subject" to message.subject,
                "messageBody" to message.messageBody,
                "conversationSid" to message.conversation.sid,
                "participantSid" to message.participantSid,
                "participant" to participantToMap(message.participant),
                "messageIndex" to message.messageIndex,
                "type" to message.type.toString(),
                "media" to mapMedia(message),
                "hasMedia" to message.hasMedia(),
                "attributes" to attributesToMap(message.attributes)
        )
    }

    fun participantListToPigeon(participants: List<Participant>?): List<Api.ParticipantData> {
        if (participants == null) {
            return listOf()
        }
        return participants.mapNotNull { participantToPigeon(it) }
    }

    fun participantToMap(participant: Participant?): Map<String, Any?>? {
        if (participant == null) {
            return null
        }
        return mapOf(
                "sid" to participant.sid,
                "conversationSid" to participant.conversation.sid,
                "lastReadMessageIndex" to participant.lastReadMessageIndex,
                "lastReadTimestamp" to participant.lastReadTimestamp,
                "dateCreated" to participant.dateCreated,
                "dateUpdated" to participant.dateUpdated,
                "identity" to participant.identity,
                "type" to participant.type.toString()
        )
    }

    fun participantToPigeon(participant: Participant?): Api.ParticipantData? {
        if (participant == null) {
            return null
        }
        val result = Api.ParticipantData()
                result.sid = participant.sid
        result.conversationSid = participant.conversation.sid
        result.lastReadMessageIndex = participant.lastReadMessageIndex
        result.lastReadTimestamp = participant.lastReadTimestamp
        result.dateCreated = participant.dateCreated
        result.dateUpdated = participant.dateUpdated
        result.identity = participant.identity
        result.type = participant.type.toString()
        return result
    }

    fun userToMap(user: User?): Map<String, Any>? {
        if (user == null) return null
        return mapOf(
                "friendlyName" to user.friendlyName,
                "attributes" to attributesToMap(user.attributes),
                "identity" to user.identity,
                "isOnline" to user.isOnline,
                "isNotifiable" to user.isNotifiable,
                "isSubscribed" to user.isSubscribed
        )
    }

    fun userToPigeon(user: User?): Api.UserData? {
        if (user == null) return null
        val result = Api.UserData()

        result.friendlyName = user.friendlyName
        result.attributes = attributesToPigeon(user.attributes)
        result.identity = user.identity
        result.isOnline = user.isOnline
        result.isNotifiable = user.isNotifiable
        result.isSubscribed = user.isSubscribed
        return result
    }

    private fun dateToString(date: Date?): String? {
        if (date == null) return null
        val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss ZZZ")
        return dateFormat.format(date)
    }

    private fun mapMedia(message: Message): Map<String, Any?>? {
        if (!message.hasMedia()) {
            return null
        }

        return mapOf(
                "sid" to message.mediaSid,
                "fileName" to message.mediaFileName,
                "type" to message.mediaType,
                "size" to message.mediaSize,
                "conversationSid" to message.conversationSid,
                "messageIndex" to message.messageIndex,
                "messageSid" to message.sid
        )
    }

    fun mediaToPigeon(message: Message): Api.MessageMediaData? {
        if (!message.hasMedia()) {
            return null
        }

        val result = Api.MessageMediaData()
        result.sid = message.mediaSid
        result.fileName = message.mediaFileName
        result.type = message.mediaType
        result.size = message.mediaSize
        result.conversationSid = message.conversationSid
        result.messageIndex = message.messageIndex
        result.messageSid = message.sid
        return result
    }

    fun errorInfoToMap(errorInfo: ErrorInfo?): Map<String, Any?>? {
        errorInfo ?: return null

        return mapOf(
                "code" to errorInfo.code,
                "message" to errorInfo.message,
                "status" to errorInfo.status
        )
    }

    fun errorInfoToPigeon(errorInfo: ErrorInfo): Api.ErrorInfoData {
        val errorData = Api.ErrorInfoData()
        errorData.status = errorInfo.status.toLong()
        errorData.message = errorInfo.message
        errorData.code = errorInfo.code.toLong()
        return errorData
    }
}
